### ISSUE TITLE
Exception publishing diagnostics

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/clojure-lsp#1054]
**Describe the bug**
Occasionally the logs will show an exception while publishing diagnostics:

<details>
 <summary><b>Log - clojure-lsp</b></summary>
<pre>
2022-06-16T16:56:54.639Z  ERROR [taoensso.timbre:828] - Uncaught exception on thread: async-dispatch-8
                                              java.lang.Thread.run              Thread.java:  829
                                                               ...
clojure.core.async.impl.concurrent/counted-thread-factory/reify/fn           concurrent.clj:   29
                java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  628
                 java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1128
                                                               ...
             clojure.core.async.impl.channels.ManyToManyChannel/fn             channels.clj:  139
                       clojure.core.async.impl.ioc-macros/take!/fn           ioc_macros.clj:  991
      clojure.core.async.impl.ioc-macros/run-state-machine-wrapped           ioc_macros.clj:  982
              clojure.core.async.impl.ioc-macros/run-state-machine           ioc_macros.clj:  978
           clojure-lsp.server/run-server!/fn/state-machine--auto--               server.clj:  246
        clojure-lsp.server/run-server!/fn/state-machine--auto--/fn               server.clj:  246
          clojure-lsp.server.ClojureLspProducer/publish-diagnostic               server.clj:   86
                                                               ...
                                               clojure.core/format                 core.clj: 5763
                                               clojure.core/format                 core.clj: 5769
                                           java.lang.String.format              String.java: 2897
                                        java.util.Formatter.format           Formatter.java: 2609
                                        java.util.Formatter.format           Formatter.java: 2673
                         java.util.Formatter$FormatSpecifier.print           Formatter.java: 2902
                java.util.Formatter$FormatSpecifier.printCharacter           Formatter.java: 3014
                java.util.Formatter$FormatSpecifier.failConversion           Formatter.java: 4426
java.util.IllegalFormatConversionException: c != java.lang.String
    argumentClass: java.lang.String
       conversion: \c
</pre>
</details>

After this happens, diagnostics stop being published—the async publisher is dead. Usually, a little later, Emacs will freeze, possible because it's waiting for the server to do something that it isn't going to do.

It doesn't always complain about `\c` as a conversion. I've also seen `\d` and `\l`.

**To Reproduce**
I have a reproduction, but unfortunately, I haven't figured out exactly when the problem happens organically. It can happen any time, but seems most likely to happen during startup, when the diagnostics for all the files are published. Once its happening, I can reliably reproduce it by restarting the server, as long as I don't touch anything. But any changes can make it go away—it goes away if I restart Emacs; I think it may also go away if I do a "Save all".

My synthetic reproduction is to put a file named `%e%.clj` in the filesystem, and open it with Emacs. This causes a formatting exception, though with a slightly different message... The `%` symbols in `%e%.clj` are escaped into `%25`, and then the exception is `java.util.UnknownFormatConversionException: Conversion = '2'`. To match the real exception, somehow we'd have to get `%c` into the escaped uri.

**Expected behavior**
Diagnostics should be published, or at a minimum, the publisher thread shouldn't die.

**Additional context**
The error happens [here](https://github.com/clojure-lsp/clojure-lsp/blob/efbf686528c69e36f3c444227b8e2bbe1cc9d726/cli/src/clojure_lsp/server.clj#L82). Based on my reproduction technique, I suspect that Emacs adds `%` symbols to `(:uri diagnostics)` as its locking or auto-saving files. Then, we pass a string with extra `%` symbols in it to `log-timing`. Since `log-timing` calls `format` as well, it interprets those `%` symbols as interpolation locations.

I have some ideas for how to fix this:

1. log-timing could try to escape the messages it receives by replacing `%` with `%%`. I'm not sure that would be sufficient... I'm having trouble finding docs about how to escape strings that are going to be passed to `format`.
2. log-timing could forego `format`, and instead just tack the timing onto the plain string it receives. We'd have to remove `"took %s"` from all of the messages. This seems less error prone than 1).
3. [These](https://github.com/clojure-lsp/clojure-lsp/blob/3514f39ae502f31dec0d35d4e42259322545aca0/cli/src/clojure_lsp/server.clj#L240-L245) go-loops should catch and log errors like their counterparts, instead of crashing.

A final thought... I think, though I'm not sure, that when this error is thrown it propagates out of [the go-loop](https://github.com/clojure-lsp/clojure-lsp/blob/3514f39ae502f31dec0d35d4e42259322545aca0/cli/src/clojure_lsp/server.clj#L243-L245) to [the end of `run-server!`](https://github.com/clojure-lsp/clojure-lsp/blob/3514f39ae502f31dec0d35d4e42259322545aca0/cli/src/clojure_lsp/server.clj#L264). I'm not sure what effect exceptions in `run-server!` have on the other go-loops, on the launcher, and on the server as a whole. I imagine it might be bad. I'm not sure we can do anything to fix whatever badness happens, but we could do a best-effort to log the exception by wrapping `run-server!` in a try/catch block.

**Comment by ericdallo** [comment]
yeah, I think maybe each go-loop should have its own try/catch block

**Comment by ericdallo** [comment]
@mainej did a commit adding try catch blocks for those tasks, even so, I noticed that if an exception is thrown on changes for example, the editor will freeze sometime later, probably related to some place waiting for the `processing-changes`.

**Comment by mainej** [comment]
> the editor will freeze sometime later, probably related to some place waiting for the `processing-changes`.

Hmm... that sucks, but it makes sense. I'm not really sure what we can do about that.

I guess for this particular problem we should do 1. or 2. as well. But if the try/catch doesn't really protect us, we'll likely run into this problem again someday.

**Comment by ericdallo** [comment]
Yeah, I think we should timeout the :processing-changes value, to avoid this lock, and a not long timeout as it's pretty hard to some task take more than 5s I think

**Comment by mainej** [comment]
I still don't know what's happening in #1058, but I've figured out exactly when this bug, #1054, happens.

Emacs creates [lock files](https://www.gnu.org/software/emacs/manual/html_node/elisp/File-Locks.html) to prevent multiple Emacs sessions from editing the same file. When editing a file named `foo.clj`, it adds another file in the same directory named `#.foo.clj`. 

When clojure-lsp is restarted, that file is analyzed, linted and its diagnostics are published. Within clojure-lsp, the file's URI is something like `file:///.%23foo.clj`. The `#` is URL encoded into `%23`. 

The bug is that as the diagnostics are being published, `%23f` gets interpolated into a string, which is then run through `format`.

```clojure
(format (str "Publishing 5 diagnostics for " "file:///.%23foo.clj" " took %s")
        "35ms")
```
And that throws an error:
```
Execution error (IllegalFormatConversionException) at java.util.Formatter$FormatSpecifier/failConversion (Formatter.java:4426).
f != java.lang.String
```

You can get different errors by changing the first character of the file's name. In this case it complains about `f` because that's the first character of `foo.clj`.

So... what should be done about this? I can configure my Emacs to put lock files somewhere else. That will fix it for me, but other people are likely to run into this too.

Looking at clojure-lsp, I can change the code that's interpolating a URI into a `shared/logging-time` message. I'll submit a PR for that.

More generally, it's definitely an error to interpolate URIs into strings that will be passed to `format`. I don't know how to prevent that, or how to truly sanitize these strings. This is a rough attempt but I suspect the problem is subtler than this:
```clojure
(format (string/replace (str "Publishing 5 diagnostics for " "file:///.%23foo.clj" " took %s")
                        #"%" "%%")
        "35ms")
;; => "Publishing 5 diagnostics for file:///.%23foo.clj took 35ms"
```
That would work in some places, but the same problem could pop up elsewhere.

Alternatively you could say the problem is in clj-kondo, because it shouldn't consider `.#foo.clj` a source file. 

Or you could say it's not a problem at all. @ericdallo FWIW, the try/catch block works OK for me. An error is printed and both the client and server keep working fine.

**Comment by ericdallo** [comment]
Oh, that makes sense now, I use doom-emacs which disable this behavior on emacs.

> Alternatively you could say the problem is in clj-kondo, because it shouldn't consider .#foo.clj a source file.

We pass the file to clj-kondo, so it's probably our fault.
I was wondering if we can just ignore those files on lsp-mode side or as last resource on clojure-lsp side? Does it make sense for lsp-mode to send those emacs files to be analyzed?

> Or you could say it's not a problem at all. @ericdallo FWIW, the try/catch block works OK for me. An error is printed and both the client and server keep working fine.

Good

**Comment by mainej** [comment]
> We pass the file to clj-kondo

At startup we only pass source _directories_ to clj-kondo, not source files. It's clj-kondo that [scans the directory](https://github.com/clj-kondo/clj-kondo/blob/478cb08282bcf43b1443ee24f0664a85bc3ac8cc/src/clj_kondo/impl/core.clj#L296) and [decides it's a source file](https://github.com/clj-kondo/clj-kondo/blob/478cb08282bcf43b1443ee24f0664a85bc3ac8cc/src/clj_kondo/impl/core.clj#L183-L188). So either clojure-lsp has to take over that job—turning a directory into a list of source files and filtering out the "bad" files—or it would have to be fixed in clj-kondo.

> ignore those files on lsp-mode

I can't remember if I've ever seen this problem after startup... I suppose if Emacs ever sends that file as a "did-open" or "did-change" we'd see it. But I don't think it does. So maybe effectively lsp-mode already ignores those files.

**Comment by ericdallo** [comment]
I see, If it doesn't happen on did-open/did-change or did-change-watched-file, it should be a kondo issue
